### PR TITLE
feat: program testing draft

### PR DESCRIPTION
**Pull request by ixmorrow**

### CODE REVIEW

File: content/program-testing.md [hunk 1]
```diff
@@ -0,0 +1,488 @@
...
+fn it_adds_two() {
+    assert_eq!(4, example_lib::add_two(2));
+}
+```
+
+Each file in the `tests` directory is a separate crate, so we will need to bring our library of code that we want to test into each file’s scope - that’s what the `use example_lib` line is doing.
+
+We don’t need to annotate the tests in the `tests` directory with `#[cfg(test)]` because Cargo will only compile files inside the `tests` directory when we run `cargo test`. Cargo is pretty smart, right?
+
+Once you have tests written (either unit, integration, or both), all you need to do is run `cargo test-bpf` or `cargo test` and they will execute. A successful completion of a single unit and single integration test will output something like this to the command line.
```
Comment (jamesrp13) [inline content/program-testing.md:69]: Is there a practical difference between `cargo test-bf` and `cargo test`?

File: content/program-testing.md [hunk 2]
```diff
@@ -0,0 +1,488 @@
...
+
+```
```
[the remark this answers is not in the document]
Comment (jamesrp13) [reply content/program-testing.md:71]: Oh nvm this is the output

File: content/program-testing.md [hunk 3]
```diff
@@ -0,0 +1,488 @@
...
+
+![local validator](../assets/local-validator.png)
+
+To interact with a running local validator, open a new terminal and configure the CLI to target your local host.
+
+```
+solana config set --url localhost
+```
+
+Once set, go ahead and check your wallet balance `solana balance` and then airdrop yourself some tokens `solana airdrop 50`. There is no limit to the amount of tokens you can airdrop yourself on localhost, unlike devnet. You can even open a third terminal to monitor logs that your program generates by running `solana logs`.
```
Comment (jamesrp13) [inline content/program-testing.md:114]: This is reading like a demo at this point. Just little things like "go ahead and do this." I'd prefer that it be "this is how you do this"

File: content/program-testing.md [hunk 4]
```diff
@@ -0,0 +1,488 @@
...
+  2 passing (1s)
+```
+
+## Debugging
+
+When testing a program, you may run into errors or output that is unexpected, that’s where debugging comes into play. Debugging refers to the process of finding and resolving errors or unintended behavior of software programs.
+
+### Error Codes
+
+When writing programs in general, it is inevitable that you will spend a good portion of your time debugging. There are program errors that you may come across from time to time and they can be a little confusing. Some program errors have a hexadecimal code associated with them that looks something along the lines of `0xa4` which might not seem to make any sense. This error code is actually a hexadecimal representation of this error’s decimal index inside the error enum of the program that returned it. So, if an error is returned in your program with a hexadecimal code and no message, try converting that hexadecimal number *x* to decimal *y* and looking up the corresponding error at *y* index of the program’s error enum.
```
Comment (jamesrp13) [inline content/program-testing.md:268]: Could we get an example here of converting the hexadecimal number and then looking up the corresponding error number?

File: content/program-testing.md [hunk 5]
```diff
@@ -0,0 +1,488 @@
...
+
+### Program Logs
+
+Another helpful tool when debugging is logging. Solana makes it very easy to create new custom logs with the `msg!()` macro and, as you’ve seen in this course, you can even log data from accounts inside the program. This is probably one of the most helpful tools when it comes to debugging because you can see exactly how your program is interacting with the accounts involved and if it’s doing what is expected by logging data throughout the program.
```
Comment (jamesrp13) [inline content/program-testing.md:272]: I found last night that I could use `println!` in the unit tests, which makes sense since it's not run on chain. It may be worth mentioning this as well.

**Review by jamesrp13 — CHANGES_REQUESTED** [review]
I've read through up to the demo (haven't gone through the demo yet). I've left comments with specific thoughts, but also have a broad request about the structure of the overview:

After seeing this, I think the top level headers in the testing area should be (with different wording):
- Unit tests
- Integration tests
- RPC tests

And then within each is a section on:
1. what it is
2. how to do it

The current structure feels a bit disjointed where we talk about all of the test types individually, then circle back and talk about each of them again. I think it would be more cohesive to do each test type in it's own ## section. Open to pushback on that but that's my two cents.

File: content/program-testing.md [hunk 6]
```diff
@@ -0,0 +1,488 @@
...
+
+### Tests in Rust
+
+The rust package manager, Cargo, natively has some tools built into it to help developers write their own automated tests. Whenever we make a new library project with Cargo, a test module with a test function in it is automatically generated for us.
+
+The Rust community thinks about tests in terms of two main categories: unit tests and integration tests. *Unit tests* are small and more focused, testing one module in isolation at a time, and can test private interfaces. *Integration tests* are entirely external to your library and use your code in the same way any other external code would, using only the public interface and potentially exercising multiple modules per test.
+
+The purpose of unit tests is to test each unit of code in isolation from the rest of the code to quickly pinpoint where code is and isn’t working as expected. Unit tests reside in the `src` directory in the file with the code they are testing. Unit tests are declared inside a module named `tests` annotated with `cfg(test)`. At its simplest, a test in Rust is a function that’s annotated with the `#[test]` attribute.
```
[the remark this answers is not in the document]
Comment (ixmorrow) [reply content/program-testing.md:33]: This section came almost directly from the rust book https://doc.rust-lang.org/stable/book/ch11-03-test-organization.html so that's what I was using as reference

File: content/program-testing.md (hunk 1 above)
Comment (ixmorrow) [reply content/program-testing.md:69]: yes, I know when testing solana programs if your program that you're testing is supposed to create a new account then `cargo test` will not work and will throw some weird error I came across. The solution was to use `cargo test-bpf` instead - I dont fully understand why just `cargo test` doesn't work in that scenario, still looking into it.

File: content/program-testing.md (hunk 5 above)
Comment (ixmorrow) [reply content/program-testing.md:272]: oh nice I'll try that out, good find!

File: content/program-testing.md [hunk 7]
```diff
@@ -0,0 +1,488 @@
...
+# TL;DR
+
+- Testing is a **key component** of smart contract development, it ensures the code works as intended before releasing it to the public.
+- Solana programs support three different types of tests (unit, integration, and client-side) all of which have a specific purpose.
+
+# Overview
+
+## Testing Concepts
+
+Testing in software is very common and there are actually entire careers fields dedicated to just creating and running tests. A robust testing process can minimize the amount of bugs developers introduce into production code by catching them before they pose a real issue. You obviously cannot test *everything*, but it’s important to try to think of all the ways you can try to break your program or cause some unintended actions with your tests. This is especially imperative when developing smart contracts because a single bug can lead to millions of dollars lost or stolen.
```
[the remark this answers is not in the document]
Comment (ixmorrow) [reply content/program-testing.md:23]: I like it, good call!

**Comment by mixelpixel** [comment]
@ixmorrow I sent you a PR with some layout fixes: https://github.com/ixmorrow/solana-course/pull/1

**Review by mixelpixel — CHANGES_REQUESTED** [review]
layout fixes https://github.com/ixmorrow/solana-course/pull/1

File: content/program-testing.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (mixelpixel) [reply content/program-testing.md:33]: @jamesrp13 is this resolved?

[the remark this answers is not in the document]
Comment (ixmorrow) [reply content/program-testing.md]: I just reworded it slightly

**Comment by ixmorrow** [comment]
> layout fixes [ixmorrow#1](https://github.com/ixmorrow/solana-course/pull/1)

merged 🥇 💯

File: content/program-testing.md [hunk 8]
```diff
@@ -0,0 +1,534 @@
...
+assert_matches!(banks_client.process_transaction(transaction).await, Ok(_));
+```
+
+You can now run this test with `cargo test-bpf` and if it’s successful, you’ll be able to see the program logs and the final test result in the terminal. It may take a while to compile and run the test. Feel free to take a look at the solution code here.
+
+Take a look at the testing script in the `ts` directory and compare it to what we just wrote in Rust. They are doing almost the exact same thing, but seeing how the code differs between Typescript and Rust can be eye-opening sometimes.
+
+# Challenge
+
+We just wrote a single unit test in Rust, but a proper testing architecture is made up of more than just one test. As a challenge, build on top of what we just did and write some unit tests in Rust (or Typescript if you've seen enough Rust for the day) that test the other instructions in the program and also think about how you can write some tests with malicious or inaccurate code that's supposed to return an error from the program.
```
Comment (mixelpixel) [inline content/program-testing.md:534]: @ixmorrow please make this at least two sentences :)  Maybe just break it in the middle after the parentheses?

File: content/program-testing.md [hunk 9]
```diff
@@ -0,0 +1,534 @@
...
+    ],
+    Some(&payer.pubkey()),
+);
+transaction.sign(&[&payer, &mint_keypair], recent_blockhash);
+
+// Process transaction and compare the result
+assert_matches!(banks_client.process_transaction(transaction).await, Ok(_));
+```
+
+You can now run this test with `cargo test-bpf` and if it’s successful, you’ll be able to see the program logs and the final test result in the terminal. It may take a while to compile and run the test. Feel free to take a look at the solution code here.
```
Comment (mixelpixel) [inline content/program-testing.md:528]: @ixmorrow fix: runon
from: You can now run this test with `cargo test-bpf` and if it’s successful,

to: You can now run this test with `cargo test-bpf`. If it’s successful, ...

File: content/program-testing.md [hunk 10]
```diff
@@ -0,0 +1,534 @@
...
+    &mint_auth,
+    Some(&mint_auth),
+    9,
+)
+.unwrap();
+```
+
+We use two functions that we've imported from Solana crates to help create the `create_account` and `initialize_mint` instructions.
+
+Next, we need to create/derive the review, comment counter, and user associated token account addresses.
```
Comment (mixelpixel) [inline content/program-testing.md:453]: @ixmorrow suggestion: 
from: Next, we need to create/derive the review,  ...

to: Next, we need to derive the review, ...

...mostly I just dislike the undecisive verb slash. Maybe just "Next, we need to find the review, ..."?

File: content/program-testing.md [hunk 11]
```diff
@@ -0,0 +1,534 @@
...
+        .await;
+
+        ...
+    }
+}
+```
+
+### 3. Construct transaction
+
+Our first unit test will test whether or not our program is actually functioning as intended, to test this we’ll need to create a transaction to submit to the program. We'll be making use of the `solana_sdk` crate to help us do this, it may help to think about the steps you would need to go through to build this transaction from a typescript client when doing this.
```
Comment (mixelpixel) [inline content/program-testing.md:419]: @ixmorrow fix: first sentence run on

suggestion:

> Our first unit test will test whether or not our program is actually functioning as intended. To test this we’ll need to create a transaction to submit to the program. ...

File: content/program-testing.md [hunk 12]
```diff
@@ -0,0 +1,534 @@
...
+
+Since we have basically been testing programs via a client or script for this whole course, this demo is going to focus on writing a unit tests in Rust. We'll be writing some tests for the Movie Review program that we've been working on. If you've been following along then you probably already have the starter code, as we'll just be picking up where we left off with the CPI lesson.
+
+### 1. Clone starter code
+
+No worries if you don't already have the code locally, you can [clone the starter code from Github](https://github.com/ixmorrow/movie-review-tokens).
+
+### 2. Initialize testing framework
+
+Now we’re going to focus on writing some unit tests for this program. Our tests will focus on whether or not the program works as intended when provided the proper data, as well as how it handles unexpected or malicious input. Remember, our goal when testing is to try to catch bugs and ensure security so it’s important to also write tests that are *supposed* to fail. We’re not just focused on testing if the code works, we’re also interested in testing the robustness of our code.
```
Comment (mixelpixel) [inline content/program-testing.md:353]: @ixmorrow fix: 2nd sentence runon

suggestion:
> Our tests will focus on whether or not the program works as intended when provided the proper data. We'll also test how it handles unexpected or malicious input.

File: content/program-testing.md [hunk 13]
```diff
@@ -0,0 +1,534 @@
...
+```rust
+let authority_pubkey = Box::new(Pubkey::create_program_address(authority_signer_seeds, program_id)?);
+
+if *authority_pubkey != *authority_info.key {
+      msg!("Derived lending market authority {} does not match the lending market authority provided {}");
+      return Err();
+}
+```
+
+You simply wrap whatever variables you'd like to remove from the stack in the `Box` struct, and the compiler will allocate memory on the heap and place the value of what's wrapped inside the `Box` there.
```
Comment (mixelpixel) [inline content/program-testing.md:339]: @ixmorrow fix: runon & add serial comma

sugg:
> You simply wrap whatever variables you'd like to remove from the stack in the `Box` struct. The compiler will allocate memory on the heap, and place there the value of what's wrapped inside the `Box`.

Note, I don't love the way ending with "there" or starting with "place there" read, but the latter is preferable, unless you can think of a better way to state the case?

File: content/program-testing.md [hunk 14]
```diff
@@ -0,0 +1,534 @@
...
+
+...
+}
+```
+
+For some more detailed information regarding the compute budget [check out the docs](https://docs.solana.com/developing/programming-model/runtime#compute-budget).
+
+### Stack size
+
+Every program has access to [4KB of stack frame size when executing]("https://docs.solana.com/developing/on-chain-programs/overview#stack"). This is a different concept from the compute budget we just discussed because the stack size limit is focused solely on memory, while the compute budget is meant to limit computationally intensive actions. Many programming languages don’t require you to think about the stack and the heap very often. But in a systems programming language like Rust, whether a value is on the stack or the heap can make a large difference - especially when working within a constrained environment like a blockchain. If you aren't familiar with the differences between the two, [the Rust book has a great explanation](https://doc.rust-lang.org/stable/book/ch04-01-what-is-ownership.html).
```
Comment (mixelpixel) [inline content/program-testing.md:312]: @ixmorrow just my opinion, but I love this paragraph

File: content/program-testing.md [hunk 15]
```diff
@@ -0,0 +1,534 @@
...
+
+Another helpful tool when debugging is logging. Solana makes it very easy to create new custom logs with the `msg!()` macro and, as you’ve seen in this course, you can even log data from accounts inside the program. This is probably one of the most helpful tools when it comes to debugging because you can see exactly how your program is interacting with the accounts involved and if it’s doing what is expected by logging data throughout the program.
+
+When writing unit tests in Rust, you cannot use the `msg!()` macro to log information within the test itself, instead you'll have to use the Rust native `println!()` macro. `msg!()` statements inside the program code will still work, you just can't log within the test with it.
+
+### Compute budget
+
+Developing on a blockchain comes with some unique constraints, one of those on Solana is the compute budget. All Solana transactions are restricted to a per instruction compute budget (with plans of moving this to a per transaction basis), the compute budget is meant to prevent a program from abusing resources. Every instruction has a budget of 200,000 compute units and different actions consume different amounts of compute units.
```
Comment (mixelpixel) [inline content/program-testing.md:289]: @ixmorrow fix: runons

suggestions:
> Developing on a blockchain comes with some unique constraints. One of those constraints on Solana is the compute budget. All Solana transactions are restricted to a per instruction compute budget (with plans of moving this to a per transaction basis). The compute budget is meant to prevent a program from abusing resources. Every instruction has a budget of 200,000 compute units and different actions consume different amounts of compute units.

Also suggest rewording the parenthetical like so:
> Although there are plans to move to a *per transaction* basis, all Solana transactions are restricted to a *per instruction* compute budget.

File: content/program-testing.md [hunk 16]
```diff
@@ -0,0 +1,534 @@
...
+
+### Program logs
+
+Another helpful tool when debugging is logging. Solana makes it very easy to create new custom logs with the `msg!()` macro and, as you’ve seen in this course, you can even log data from accounts inside the program. This is probably one of the most helpful tools when it comes to debugging because you can see exactly how your program is interacting with the accounts involved and if it’s doing what is expected by logging data throughout the program.
+
+When writing unit tests in Rust, you cannot use the `msg!()` macro to log information within the test itself, instead you'll have to use the Rust native `println!()` macro. `msg!()` statements inside the program code will still work, you just can't log within the test with it.
```
Comment (mixelpixel) [inline content/program-testing.md:285]: @ivan fix: runon

sugg:
> When writing unit tests in Rust, you cannot use the `msg!()` macro to log information within the test itself. Instead you'll have to use the Rust native `println!()` macro. `msg!()` statements inside the program code will still work, you just can't log within the test with it.

File: content/program-testing.md [hunk 17]
```diff
@@ -0,0 +1,534 @@
...
+
+For example, if you were to receive an error sending a transaction to the SPL Token Program with the error code `0x01`, the decimal equivalent of this is 1. [Looking at the source code of the Token Program](https://github.com/solana-labs/solana-program-library/blob/master/token/program/src/error.rs), we can see that the error located at this index in the program's error enum is `InsufficientFunds`. You'll need to have access to the source code of any program that returns a custom program error code to translate it.
+
+### Program logs
+
+Another helpful tool when debugging is logging. Solana makes it very easy to create new custom logs with the `msg!()` macro and, as you’ve seen in this course, you can even log data from accounts inside the program. This is probably one of the most helpful tools when it comes to debugging because you can see exactly how your program is interacting with the accounts involved and if it’s doing what is expected by logging data throughout the program.
```
Comment (mixelpixel) [inline content/program-testing.md:283]: @ixmorrow fix: 3rd sentence runon

sugg: 
> This is probably one of the most helpful tools when it comes to debugging because you can see exactly how your program is interacting with the accounts involved. You can see if it’s doing what is expected by logging data throughout the program.

File: content/program-testing.md [hunk 18]
```diff
@@ -0,0 +1,534 @@
...
+    2 passing (1s)
+```
+
+## Debugging
+
+When testing a program, you may run into errors or output that is unexpected, that’s where debugging comes into play. Debugging refers to the process of finding and resolving errors or unintended behavior of software programs.
+
+### Error codes
+
+When writing programs in general, it is inevitable that you will spend a good portion of your time debugging. There are program errors that you may come across from time to time and they can be a little confusing. Some program errors have a hexadecimal code associated with them that looks something along the lines of `Program [program_id] failed: custom program error: 0x01` which might not seem to make any sense. This error code is actually a hexadecimal representation of this error’s decimal index inside the error enum of the program that returned it. So, if an error is returned in your program with a hexadecimal code and no message, try converting that hexadecimal number *x* to decimal *y* and looking up the corresponding error at *y* index of the program’s error enum.
```
Comment (mixelpixel) [inline content/program-testing.md:277]: @ixmorrow fix: runon

3rd sentence sugg:
> Some program errors have a hexadecimal code associated with them which might not seem to make any sense. They look something along the lines of `Program [program_id] failed: custom program error: 0x01`.

4th sentence is a littlw awkward. Sugg:
>This error code is actually a hexadecimal representation of the error’s decimal index inside the error enum of the program that returned it.

The last sentence is great - what a moutful!

...if that's factually correct?

File: content/program-testing.md [hunk 19]
```diff
@@ -0,0 +1,534 @@
...
+
+`const RPC_ENDPOINT_URL = "https://api.devnet.solana.com"`
+
+Feel free to read up on the [Solana Test Validator docs](https://docs.solana.com/developing/test-validator).
+
+### How to build RPC tests
+
+As stated before, you don’t have to write your unit tests in Rust - you can actually write them in just about any language you want by deploying the program to a cluster and submitting transactions from a client for the program to process. The most common way of conducting tests like this is by deploying to a local validator and writing a client testing script in Typescript using the [Mocha testing framework](https://mochajs.org/) paired with the [Chai assertion library](https://www.chaijs.com/) (it’s important to note that you can use just about any testing framework or any language for these tests, as long as there are Solana SDKs available to use).
+
+Install mocha and chai with `npm install mocha chai`
```
Comment (mixelpixel) [inline content/program-testing.md:201]: @ixmorrow sugg: I think we want to capitalize the names here:

> Install Mocha and Chai with `npm install mocha chai`

File: content/program-testing.md [hunk 20]
```diff
@@ -0,0 +1,534 @@
...
+
+`solana config set --url devnet`
+
+`const RPC_ENDPOINT_URL = "https://api.devnet.solana.com"`
+
+Feel free to read up on the [Solana Test Validator docs](https://docs.solana.com/developing/test-validator).
+
+### How to build RPC tests
+
+As stated before, you don’t have to write your unit tests in Rust - you can actually write them in just about any language you want by deploying the program to a cluster and submitting transactions from a client for the program to process. The most common way of conducting tests like this is by deploying to a local validator and writing a client testing script in Typescript using the [Mocha testing framework](https://mochajs.org/) paired with the [Chai assertion library](https://www.chaijs.com/) (it’s important to note that you can use just about any testing framework or any language for these tests, as long as there are Solana SDKs available to use).
```
Comment (mixelpixel) [inline content/program-testing.md:199]: @ixmorrow sugg: I don't think last sentence needs to be a parenthetical.

maybe something a la
> While those are the most common right now, it’s important to note that you can use just about any testing framework or any language for these tests as long as there are Solana SDKs available to use!

File: content/program-testing.md [hunk 21]
```diff
@@ -0,0 +1,534 @@
...
+Account data resizing not supported yet: 0 -> 1000. Consider making this test conditional on `#[cfg(feature = "test-bpf")]`
+```
+
+If you come across this, try using `cargo test-bpf` instead.
+
+## RPC tests
+
+### What are RPC tests?
+
+Solana programs support unit and integration tests in Rust like we just discussed, but there is also a third way to test your programs that is unique to smart contract development. The alternative method is to test your program is by deploying it to either Devnet or a local validator and sending transactions to it from some client that you created. Deploying to Devnet and then sending transactions to the program is essentially what we have been doing for the entirety of this course so you should be pretty familiar with that already and the first lesson of this module covered deploying to a local validator.
```
Comment (mixelpixel) [inline content/program-testing.md:163]: @ixmorrow fix: last sentence runon

sugg: 
> Deploying to Devnet and then sending transactions to the program is essentially what we have been doing for the entirety of this course so you should be pretty familiar with that already. If you'd like a refresher, review the first lesson of this module which covered deploying to a local validator.

File: content/program-testing.md [hunk 22]
```diff
@@ -0,0 +1,534 @@
...
+
+#[test]
+fn it_adds_two() {
+    assert_eq!(4, example_lib::add_two(2));
+}
+```
+
+Each file in the `tests` directory is a separate crate, so we will need to bring our library of code that we want to test into each file’s scope - that’s what the `use example_lib` line is doing.
+
+We don’t need to annotate the tests in the `tests` directory with `#[cfg(test)]` because Cargo will only compile files inside the `tests` directory when we run `cargo test`. Cargo is pretty smart, right?
```
Comment (mixelpixel) [inline content/program-testing.md:120]: @ixmorrow nice

File: content/program-testing.md [hunk 23]
```diff
@@ -0,0 +1,534 @@
...
+mod tests {
+    #[test]
+    fn it_works() {
+        let result = 2 + 2;
+        assert_eq!(result, 4);
+    }
+}
+```
+
+The `cfg` attribute stands for *configuration* and tells Rust that the following item should only be included given a certain configuration option. In this case, the `#[cfg(test)]` annotation tells Cargo to compile our test code only if we actively run the tests with `cargo test`, this way the testing code is not run when you call `cargo build` which saves on compile time.
```
Comment (mixelpixel) [inline content/program-testing.md:44]: @ixmorrow fix: runon 2nd sentence

sugg: 
> In this case, the `#[cfg(test)]` annotation tells Cargo to compile our test code only if we actively run the tests with `cargo test`. This way the testing code is not run when you call `cargo build` and that saves on compile time.

File: content/program-testing.md [hunk 24]
```diff
@@ -0,0 +1,534 @@
...
+
+# Overview
+
+Testing in software is very common and there are actually entire careers fields dedicated to just creating and running tests. A robust testing process can minimize the amount of bugs developers introduce into production code by catching them before they pose a real issue. You obviously cannot test *everything*, but it’s important to try to think of all the ways you can try to break your program or cause some unintended actions with your tests. This is especially imperative when developing smart contracts because a single bug can lead to millions of dollars lost or stolen.
+
+Think back to what was discussed in the [Basic Security](./program-security.md) lesson, how can we write tests that determine if the security checks we've implemented actually work as intended and are sufficient? Program security and testing go hand in hand. To write good tests, it's helpful to think like an attacker. To reiterate some of what was discussed in that lesson, the goal is to not just "get the code working", but to ensure it works properly and is robust enough to properly handle malicious input.
```
Comment (mixelpixel) [inline content/program-testing.md:20]: @ixmorrow I think the question can be its own sentence

e.g.
> Think back to what was discussed in the [Basic Security](./program-security.md) lesson. How can we write tests that determine if the security checks we've implemented actually work as intended and are sufficient?

File: content/program-testing.md [hunk 25]
```diff
@@ -0,0 +1,534 @@
...
+- Debug Solana programs
+
+# TL;DR
+
+- Testing is a **key component** of smart contract development, it ensures the code works as intended before releasing it to the public.
+- Solana programs support three different types of tests (unit, integration, and client-side) all of which have a specific purpose.
+
+# Overview
+
+Testing in software is very common and there are actually entire careers fields dedicated to just creating and running tests. A robust testing process can minimize the amount of bugs developers introduce into production code by catching them before they pose a real issue. You obviously cannot test *everything*, but it’s important to try to think of all the ways you can try to break your program or cause some unintended actions with your tests. This is especially imperative when developing smart contracts because a single bug can lead to millions of dollars lost or stolen.
```
Comment (mixelpixel) [inline content/program-testing.md:18]: @ixmorrow FIX: spelling

careers fields --> career-fields

File: content/program-testing.md [hunk 26]
```diff
@@ -0,0 +1,534 @@
...
+
+*By the end of this lesson, you will be able to:*
+
+- Describe various ways to test Solana programs
+- Explain the difference between unit tests, integration tests, and RPC tests
+- Debug Solana programs
+
+# TL;DR
+
+- Testing is a **key component** of smart contract development, it ensures the code works as intended before releasing it to the public.
```
Comment (mixelpixel) [inline content/program-testing.md:13]: @ixmorrow awk

sugg: 
> - Testing is a **key component** of smart contract development because it ensures the code works as intended before releasing it to the public.

File: content/program-testing.md [hunk 27]
```diff
@@ -0,0 +1,534 @@
...
+    ...
+
+})
+```
+
+Inside the `describe` section, each test is designated with `it`
+
+```tsx
+describe("begin tests", async () => {
+    // first Mocha test
```
Comment (mixelpixel) [inline content/program-testing.md:229]: @ixmorrow FIX: casing

> // First Mocha test

File: content/program-testing.md (hunk 26 above)
Comment (mixelpixel) [reply content/program-testing.md:13]: @ixmorrow just checking in - can you resolve these comments, suggestions and notes on this PR/103?

**Comment by ixmorrow** [comment]
@mixelpixel sorry for the delay on these, just implemented your edits!

File: content/program-testing.md [hunk 28]
```diff
@@ -0,0 +1,513 @@
...
+*By the end of this lesson, you will be able to:*
+
+- Describe various ways to test Solana programs
+- Explain the difference between unit tests and integration tests
+- Debug Solana programs
+
+# TL;DR
+
+- Testing is a **key component** of smart contract development because it ensures the code works as intended before releasing it to the public.
+- Solana programs support all different types or tests, the most common being unit and integration tests.
```
Comment (mixelpixel) [inline content/program-testing.md:14]: @ixmorrow fix: or --> of

File: content/program-testing.md [hunk 29]
```diff
@@ -0,0 +1,513 @@
...
+    let program_id = Pubkey::new_unique();
+    let (mut banks_client, payer, recent_blockhash) = ProgramTest::new(
+        "pda_local",
+        program_id,
+        processor!(process_instruction),
+    )
+    .start()
+    .await;
+
+    // call helper function
```
Comment (mixelpixel) [inline content/program-testing.md:399]: call --> Call

In general, code comments get sentence casing

**Review by mixelpixel — CHANGES_REQUESTED** [review]
@ixmorrow two quick fixes - see comments on lines 14 & 399

File: content/program-testing.md [hunk 30]
```diff
@@ -0,0 +1,513 @@
...
+        assert_matches!(banks_client.process_transaction(transaction).await, Ok(_);
+    }
+}
+```
+
+In the code snippet, we created a public key to use as our `program_id` and then initialized a `ProgramTest`. Then, we create a second `Keypair` and built our `Transaction` with the appropriate parameters. Finally, we used the `banks_client` that was returned when calling `ProgramTest::new` to process this transaction and check that the return value is equal to `Ok(_)`. This is a very simple test, but from the code snippet you can see how you would go about creating more complex tests that involve more accounts and data similar to how you would client side.
+
+## Integration tests
+
+## What are integration tests?
```
Comment (mixelpixel) [inline content/program-testing.md:102]: @ixmorrow this seems like an odd choice - isn't the question a sub-section within the "Integration Tests" header?

I recommend changing it back to an h3 `###`

File: content/program-testing.md [hunk 31]
```diff
@@ -0,0 +1,513 @@
...
+
+Integration tests on the other hand, are meant to be entirely external to the code they are testing. These tests are meant to interact with your code via its public interface in the manner that it’s intended to be accessed by others. Their purpose is to test whether many parts of your library work together correctly. Units of code that work correctly on their own could have problems when integrated, so test coverage of the integrated code is important as well.
+
+### How to build integration tests with Rust
```
Comment (mixelpixel) [inline content/program-testing.md:106]: @ixmorrow nice!

File: content/program-testing.md [hunk 32]
```diff
@@ -0,0 +1,514 @@
...
+
+Tests are defined in the `tests` module with the `#[test]` attribute. When running `cargo test`, every function inside this module marked as a test will be run. You can also create helper functions that are not tests in the module, just don’t annotate them with the `#[test]` attribute.
+
+### How to build unit tests
+
+To build unit tests in Rust, you will have to use the [`solana_sdk`](https://docs.rs/solana-sdk/latest/solana_sdk/) crate. This crate is essentially the same thing as the `@solana/web3.js` package that we’ve been using in Typescript and gives us a way to interact with Solana programs in Rust. Another crate that will be useful and was made specifically for testing Solana programs is [`solana_program_test`](https://docs.rs/solana-program-test/latest/solana_program_test/#) which contains a BanksClient-based testing framework.
```
[the remark this answers is not in the document]
Comment (ixmorrow) [reply content/program-testing.md:50]: I added this line in the lesson 
"The `banks_client` returned from the `ProgramTest` will act as our interface into the testing environment"

But I have not been able to find a more in depth explanation of what the banks_client is, the Rust docs normally describe everything pretty well but there is not anything for this unfortunately. https://docs.rs/solana-program-test/latest/solana_program_test/struct.BanksClient.html

File: content/program-testing.md [hunk 33]
```diff
@@ -0,0 +1,514 @@
...
+
+The three sections of output include the unit tests, the integration test, and the doc tests.
+
+As a side note, if the program you are testing creates a new account at all (i.e. creates a PDA), then the command `cargo test` will not work and the program will fail to complete with an error like the following.
+
+```rust
+Account data resizing not supported yet: 0 -> 1000. Consider making this test conditional on `#[cfg(feature = "test-bpf")]`
+```
+
+If you come across this, try using `cargo test-bpf` instead.
```
[the remark this answers is not in the document]
Comment (ixmorrow) [reply content/program-testing.md:158]: To avoid getting issues about getting errors running `cargo test` I just changed all the references to only `cargo test-bpf`
